Add wazuh (#318)
stackhpc/stackhpc-kayobe-config: doc/source/configuration/wazuh.rst

Deploying Wazuh Manager services
================================

Setup
-----

To install specific version modify wazuh-ansible entry in ``etc/kayobe/ansible/requirements.yml``:

.. code-block:: console

  roles:
    - name: wazuh-ansible
      src: https://github.com/stackhpc/wazuh-ansible
      version: stackhpc

Version above was tested and verified, but there is no reason to use not different one.

Install the role:

``kayobe control host bootstrap``

Edit the playbook and variables to your needs:

Wazuh manager configuration
---------------------------

Wazuh manager playbook is located in ``etc/kayobe/ansible/wazuh-manager.yml``.
Running this playbook will:

* generate certificates for wazuh-manager
* setup and deploy filebeat on wazuh-manager vm
* setup and deploy wazuh-indexer on wazuh-manager vm
* setup and deploy wazuh-manager on wazuh-manager vm
* setup and deploy wazuh-dashboard on wazuh-manager vm
* copy certificates over to wazuh-manager vm

Wazuh manager variables file is located in ``etc/kayobe/inventory/group_vars/wazuh-manager/wazuh-manager``.

You may need to modify some of the variables, including:

* wazuh_manager_ip

.. note::

    NOTE:

>     If you are using multiple environments you will need to move variable files to pertinent directory,

Reviewer: I disagree with this. Moving the file will cause a diff with the base config that will make it harder to maintain, and lead to merge conflicts. It would be better to override variables where necessary, but ideally the defaults should work without changes in most cases.

Author: I have changed the note, with those suggestions in mind.

>     for example `~/deployment/src/kayobe-config/etc/kayobe/environments/production/inventory/group_vars/`

Reviewer: ```suggestion
    for example `etc/kayobe/environments/production/inventory/group_vars/`
```

Author: Fixed.

Wazuh secrets template is located in ``etc/kayobe/ansible/templates/wazuh-secrets.yml.j2``.
It will be used by wazuh secrets playbook to generate wazuh secrets vault file.

.. code-block:: console

  kayobe playbook run $KAYOBE_CONFIG_PATH/ansible/wazuh-secrets.yml
  ansible-vault encrypt --vault-password-file ~/vault.pass $KAYOBE_CONFIG_PATH/inventory/group_vars/wazuh/wazuh-manager/wazuh-secrets

TLS (optional)
--------------

You can generate your own TLS certificates, otherwise skip this section.
By default, Wazuh Ansible uses `wazuh-cert-tool.sh <https://documentation.wazuh.com/current/user-manual/certificates.html>`__
to automatically
generate certificates for wazuh-indexer (previously Elasticsearch and opendistro)
and wazuh-dashbooard (previously Kibana) using a local CA.
If the certificates directory ``etc/kayobe/ansible/wazuh/certificates``
does not exist, it will generate the following certificates in ``etc/kayobe/ansible/wazuh/certificates/certs/``
(here os-wazuh is set as ``elasticsearch_node_name`` and ``kibana_node_name``:

* Admin certificate for opendistro security
   * admin-key.pem,  admin.pem
* Node certificate
   * os-wazuh-key.pem,  os-wazuh.pem

> * HTTP certificate for Kibana (port 5601) & Elasticsearch (port 9200)

Reviewer: ```suggestion
* HTTP certificate for wazuh-dashboard (port 5601) & wazuh-indexer (port 9200)
```

Those certificates can be uploaded to ``etc/kayobe/ansible/wazuh/custom_certificates``,
and will replace certificates generated by wazuh.
Certificates should have the same name scheme as those generated by wazuh (typicaly <node-name>.pem)
The key for the external certificate should be in PKCS#8 format
(in its header it may have BEGIN PRIVATE KEY instead of BEGIN RSA PRIVATE KEY or BEGIN OPENSSH PRIVATE KEY).

Example OpenSSL rune to convert to PKCS#8:

``openssl pkcs8 -topk8 -nocrypt -in wazuh.key -out wazuh.key.pkcs8``

TODO: document how to use a local certificate. Do we need to override all certificates?

Deploy
------

Deploy Wazuh manager:

``kayobe playbook run $KAYOBE_CONFIG_PATH/ansible/wazuh-manager.yml``

If you are using the wazuh generated certificates,
this will result in the creation of some certificates and keys (in case of custom certs adjust path to it).
Encrypt the keys (and remember to commit to git):

``ansible-vault encrypt --vault-password-file ~/vault.pass $KAYOBE_CONFIG_PATH/ansible/wazuh/certificates/certs/*.key``

Verification
==============

The Wazuh portal should be accessible on port 443 of the Wazuh

> manager’s IPs (using HTTPS, with the root CA cert in ``etc/kayobe/ansible/wazuh/certificates/root-ca.pem``).

Reviewer: Is it in etc/kayobe/ansible/wazuh/certificates/certs/root-ca.pem?

Author: Fixed

=====
Wazuh
=====

Wazuh Manager
=============

Provision using infra-vms
-------------------------

Provisioning an infra VM for Wazuh Manager.

> From Xena, Kayobe supports `provisioning infra VMs <https://docs.openstack.org/kayobe/latest/deployment.html#infrastructure-vms>`__ . The StackHPC fork of Kayobe has backported this to Wallaby.

Reviewer: ```suggestion
From Xena, Kayobe supports :kayobe-doc:`provisioning infra VMs <deployment.html#infrastructure-vms>`. The StackHPC fork of Kayobe has backported this to Wallaby.
```

> The following configuration may be used as a guide. Config for infra VMs is documented `here <https://docs.openstack.org/kayobe/latest/configuration/reference/infra-vms.html>`__ .

Reviewer: ```suggestion
The following configuration may be used as a guide. Config for infra VMs is documented :kayobe-doc:`here <configuration/reference/infra-vms>`.
```

      lvnames:
        - lvname: "data"
          size: "100%VG"
          filesystem: "ext4"
          mount: true
          mntp: “/var/lib/elasticsearch”
          create: true

Define network interfaces ``etc/kayobe/inventory/group_vars/wazuh-manager/network-interfaces``:

(The following is an example - the names will depend on your particular network configuration.)

.. code-block:: console

  ---
  # Overcloud provisioning network IP information.
  provision_oc_net_interface: "ens3"

The Wazuh manager may need to be exposed externally, in which case it may require another interface.
This can be done as follows in ``etc/kayobe/inventory/group_vars/wazuh-manager/network-interfaces`` ,
with the network defined in network.yml as usual.

.. code-block:: console

  infra_vm_extra_network_interfaces:
    - "extra_net"

  # External network connectivity on ens2
  extra_net_interface: "ens2"

Add group mappings to the inventory ``etc/kayobe/inventory/groups``:

.. code-block:: console

  # Infra VM groups.

  [hypervisors:children]
  # Group that contains all hypervisors used for infra VMs
  seed-hypervisor

  [infra-vms:children]
  wazuh-manager

  [wazuh:children]

>   wazuh-manager

Reviewer: Can be removed

Author: Done.

  [wazuh-manager]
  # Empty group to provide declaration of wazuh-manager group.

Add the wazuh manager VM to the inventory ``etc/kayobe/inventory/hosts``:

.. code-block:: console

  [wazuh-manager]
  os-wazuh

> Follow the Kayobe instructions to `provision the VM <https://docs.openstack.org/kayobe/latest/deployment.html#infrastructure-vms>`__ and configure the host.

Reviewer: ```suggestion
Follow the Kayobe instructions to :kayobe-doc:`provision the VM <deployment.html#infrastructure-vms>` and configure the host.
```

VM sizing
---------

.. code-block:: console

  ---
  # Memory in MB.
  memory_mb: 16384

  # Number of vCPUs.
  vcpus: 8

  # Capacity of the infra VM data volume.
  capacity: "200G"

.. note::

    NOTE:

>     Logs will be stored in /var/ossec/ so it's a good idea to make it an LVM filesystem to make it futureproof.

Reviewer: Is this used? Seems obsolete with infra VMs

Author: You are right. Removed.

Network Setup
-------------

Your wazuh-manager VM needs to have network connection with servers which will have the wazuh-agent installed, preferably it should be in the `provision_oc_net`.

Add to ``etc/kayobe/network-allocation.yml``:

.. code-block:: console

  provision_oc_net_ips:
    <wazuh.vm.hostname>: <wazuh.vm.ip>

> 

Reviewer: This gets autogenerated when using an infra VM

Author: Yes. You are right.

|                 | 55000     | TCP            | Wazuh server RESTful API                       |
+-----------------+-----------+----------------+------------------------------------------------+
|                 | 9200      | TCP            | Wazuh indexer RESTful API                      |
+ Wazuh indexer   +-----------+----------------+------------------------------------------------+
|                 | 9300-9400 | TCP            | Wazuh indexer cluster communication            |
+-----------------+-----------+----------------+------------------------------------------------+
| Wazuh dashboard | 443       | TCP            | Wazuh web user interface                       |
+-----------------+-----------+----------------+------------------------------------------------+

Make sure group mappings for wazuh-manager are added to the inventory ``etc/kayobe/inventory/groups``:

.. code-block:: console

  # Infra VM groups.

  [wazuh:children]
  wazuh-manager

  [wazuh-manager]
  # Empty group to provide declaration of wazuh-manager group.

Add hosts group mappings to the inventory ``etc/kayobe/inventory/hosts``:

.. code-block:: console

  [wazuh-manager]
    <wazuh.vm.name>

> 

Reviewer: Did this already above

Author: Removed.

Manually provisioned VM
-----------------------

> In case where you can’t use infra-vms to deploy your wazuh-manager but you want to configure

Reviewer: ```suggestion
In case where you can’t use infra-vms to deploy your wazuh-manager VM but you want to configure
```

host using kayobe, there are some tips (note that depending on your setup this don’t have to always apply):

> * Depending on preferences host have to be part of some group in inventory. ``infra vm`` group still seems as best choice

Reviewer: ```suggestion
* Depending on preferences host have to be part of some group in inventory. ``infra-vms`` group still seems as best choice
```

  You can use ``kayobe infra vm host configure`` to configure host in this case.
  Bellow tips are based on assumption that infra-vm will be used.
* user ``stack`` with password less sudo and accessible with ssh keys needs to be present on host.
  It can be achieved in many different ways, depending on your setup.
* lvm configuration should be placed in ``host_vars/<host_name>``
* wazuh-manager host have to be part of ``infra-vms`` group (directly or as child)
* network used on host needs to be defined in ``networks.yml`` and
  if you have pre-alocated IP, it should be added to ``network-allocation.yml``. For example:

``networks.yml``:

.. code-block:: console

    undercloud_admin_net_cidr: 10.10.224.0/24
    undercloud_admin_net_allocation_pool_start: 10.10.224.3
    undercloud_admin_net_allocation_pool_end: 10.10.224.200
    undercloud_admin_net_gateway: 10.10.224.254

``network-allocation.yml``:

.. code-block:: console

    undercloud_admin_net_ips:

>       nesmetprd01: 10.10.224.5

Reviewer: provision_oc_net to be consistent with above examples?

Author: The point was to demonstrate that this network isn't one of the "standard" kayo networks. I will add clarification.

Deploying Wazuh Manager services
================================

Setup

> ================================

Reviewer: ```suggestion
-----
```

To install specific version modify wazuh-ansible entry in ``etc/kayobe/ansible/requirements.yml``:

.. code-block:: console

  roles:
    - name: wazuh-ansible
      src: https://github.com/stackhpc/wazuh-ansible
      version: stackhpc

Version above was tested and verified, but there is no reason to use not different one.

Install the role:

``kayobe control host bootstrap``

You might wish to add the following to .gitignore in kayobe-config:

.. code-block:: console

  # Wazuh
  etc/kayobe/ansible/roles/wazuh-ansible/
  etc/kayobe/ansible/vars/certificates/*

>   etc/kayobe/ansible/vars/certificates/custom_certificates/*

Reviewer: Should we add these as standard?

Author: Added. I will delete this section.

Edit the playbook and variables to your needs:

Wazuh manager configuration
---------------------------

Wazuh manager playbook is located in ``etc/kayobe/ansible/wazuh-manager.yml``.
Running this playbook will:

* generate certificates for wazuh-manager
* setup and deploy filebeat on wazuh-manager vm
* setup and deploy wazuh-indexer on wazuh-manager vm
* setup and deploy wazuh-manager on wazuh-manager vm
* setup and deploy wazuh-dashboard on wazuh-manager vm
* copy certificates over to wazuh-manager vm

Wazuh manager variables file is located in ``etc/kayobe/inventory/group_vars/wazuh-manager/wazuh-manager``.

You may need to modify some of the variables, including:

* domain_name
* wazuh_manager_ip

> * private_ip

Reviewer: I don't see this one in the config

Author: Nice catch. Deleted along with domain_name

.. note::

    NOTE:
    If you are using multiple environments you will need to move variable files to pertinent directory,
    for example `~/deployment/src/kayobe-config/etc/kayobe/environments/production/inventory/group_vars/`
    Files to be moved:
    - etc/kayobe/inventory/group_vars/wazuh/wazuh-manager/wazuh-manager
    - inventory/group_vars/wazuh/wazuh-manager/wazuh-secrets

>     - etc/kayobe/inventory/group_vars/wazuh/wazuh-agent/wazuh-agent

Reviewer: Do they need to be moved, or do you just need to set environment-specific variables in the environment inventory?

Author: I would leave this to operators to decide. Some would like to have whole config file in one place to avoid confusion, some would like file with overrides. But it also depends on standard/assumption/good practice used in this repo, so I am open to suggestions.

Author: I have also noticed that `- inventory/group_vars/wazuh/wazuh-manager/wazuh-secrets` is no more. Replaced with `etc/wazuh-manager.yml`

Secrets

> =======

Reviewer: ```suggestion
------
```

Wazuh secrets playbook is located in ``etc/kayobe/ansible/wazuh-secrets.yml``.
Running this playbook will generate and put pertinent security items into secrets

> vault file which will be placed in ``$KAYOBE_CONFIG_PATH/wazuh-secrets.yml``.

Reviewer: If using multiple environments it ends up in etc/kayobe/environments/<env>/wazuh-secrets.yml

Author: Added this information.

Remember to encrypt!

Wazuh secrets template is located in ``etc/kayobe/ansible/templates/wazuh-secrets.yml.j2``.
It will be used by wazuh secrets playbook to generate wazuh secrets vault file.

.. code-block:: console

>   kayobe playbook run $KAYOBE_CONFIG_PATH/ansible/wazuh-secrets.yml -e wazuh_user_pass=$(uuidgen) -e wazuh_admin_pass=$(uuidgen)

Reviewer: I don't think those vars are being used.
```suggestion
  kayobe playbook run $KAYOBE_CONFIG_PATH/ansible/wazuh-secrets.yml
```

  ansible-vault encrypt --vault-password-file ~/vault.pass $KAYOBE_CONFIG_PATH/inventory/group_vars/wazuh/wazuh-manager/wazuh-secrets

> ==============
> TLS (optional)
> ==============

Reviewer: ```suggestion
TLS (optional)
--------------
```

You can generate your own TLS certificates, otherwise skip this section.
By default, Wazuh Ansible uses `wazuh-cert-tool.sh <https://documentation.wazuh.com/current/user-manual/certificates.html>`__
to automatically
generate certificates for wazuh-indexer (previously Elasticsearch and opendistro)
and wazuh-dashbooard (previously Kibana) using a local CA.
If the certificates directory ``etc/kayobe/ansible/wazuh/certificates``
does not exist, it will generate the following certificates in ``etc/kayobe/ansible/wazuh/certificates/certs/``
(here os-wazuh is set as ``elasticsearch_node_name`` and ``kibana_node_name``:

> * Admin certificate for opendistro security
>    * admin.key,  admin.pem
> * Node certificate
>    * os-wazuh.key,  os-wazuh.pem
> * HTTP certificate for Kibana (port 5601) & Elasticsearch (port 9200)
>    * os-wazuh_http.key, os-wazuh_http.pem
> * Root CA certificate
>    * root-ca.key  root-ca.pem

Reviewer 2: It appears the [indexer role](https://github.com/stackhpc/wazuh-ansible/blob/970263ae2e078c63d71b9e89643fbb231c754d5a/roles/wazuh/wazuh-indexer/tasks/security_actions.yml#L29-L34) has renamed some of these keys for example `admin.key` is `admin-key.pem`

```suggestion
* Admin certificate for opendistro security
   * admin-key.pem,  admin.pem
* Node certificate
   * os-wazuh-key.pem,  os-wazuh.pem
* HTTP certificate for Kibana (port 5601) & Elasticsearch (port 9200)
   * os-wazuh_http.key, os-wazuh_http.pem
* Root CA certificate
   * root-ca.key  root-ca.pem
```

Author: Fixed. Thanks for commit suggestion!

It is also possible to use externally generated certificates for wazuh-dashboard. root-ca.pem should contain the CA chain.
Those certificates can be uploaded to ``etc/kayobe/ansible/wazuh/custom_certificates``,
and will replace certificates generated by wazuh.
Certificates should have the same name scheme as those generated by wazuh (typicaly <node-name>.pem)
The key for the external certificate should be in PKCS#8 format
(in its header it may have BEGIN PRIVATE KEY instead of BEGIN RSA PRIVATE KEY or BEGIN OPENSSH PRIVATE KEY).

Example OpenSSL rune to convert to PKCS#8:

``openssl pkcs8 -topk8 -nocrypt -in wazuh.key -out wazuh.key.pkcs8``

TODO: document how to use a local certificate. Do we need to override all certificates?

> =======
> Deploy
> =======

Reviewer: ```suggestion
Deploy
------
```

Deploy Wazuh manager:

``kayobe playbook run $KAYOBE_CONFIG_PATH/ansible/wazuh-manager.yml``

If you are using the wazuh generated certificates,
this will result in the creation of some certificates and keys (in case of custom certs adjust path to it).
Encrypt the keys (and remember to commit to git):

``ansible-vault encrypt --vault-password-file ~/vault.pass $KAYOBE_CONFIG_PATH/ansible/wazuh/certificates/certs/*.key``

Verification
==============

The Wazuh portal should be accessible on port 443 of the Wazuh

> manager’s IPs (using HTTPS, with the root CA cert in ``etc/kayobe/ansible/qwazuh/certificates/root-ca.pem``).

Reviewer: ```suggestion
manager’s IPs (using HTTPS, with the root CA cert in ``etc/kayobe/ansible/wazuh/certificates/root-ca.pem``).
```

The first login should be as the admin user,
with the opendistro_admin_password password in ``$KAYOBE_CONFIG_PATH/wazuh-secrets.yml``.
This will create the necessary indices.

Troubleshooting

Logs are in ``/var/log/wazuh-indexer/wazuh.log``. There are also logs in the journal.

============
Wazuh agents
============

Make sure group mappings for wazuh-agent are added to the inventory ``etc/kayobe/inventory/groups``:

.. code-block:: console

  [wazuh-agent:children]
  seed

>   overcloud

Reviewer: Already in groups

Author: Deleted.

  [wazuh:children]

>   wazuh-agent

Reviewer: Let's add this to groups by default

Author: Added.

Wazuh agent playbook is located in ``etc/kayobe/ansible/wazuh-agent.yml``.

Wazuh agent variables file is located in ``etc/kayobe/inventory/group_vars/wazuh-agent/wazuh-agent``.

You may need to modify some variables, including:

* wazuh_manager_address

Deploy the Wazuh agents:

``kayobe playbook run $KAYOBE_CONFIG_PATH/ansible/wazuh-agent.yml``

Verification
=============

The Wazuh agents should register with the Wazuh manager. This can be verified via the agents page in Wazuh Portal.
Check CIS benchmark output in agent section.

Additional resources:

> ---------------------

Reviewer: ```suggestion
=================
```

> For times when you need to upgrade wazuh with elasticsearch to version with opensearch or you just need to deinstall all wazuh compotnents:

Reviewer 3: *components

Author: Fixed

> Wazuh purge script: https://github.com/g0rgamesh/wazuh-server-purge

Reviewer 3: Should we at least move that from private space?

Reviewer 4: I can recreate that in stackhpc organization, I wasn't sure if we can put stuff willy-nilly into the org.

Reviewer 4: Done

==============

The Wazuh portal should be accessible on port 443 of the Wazuh

> manager’s IPs (using HTTPS, with the root CA cert in ``etc/kayobe/ansible/wazuh/certificates/wazuh-certificates/root-ca.pem``).

Reviewer: Is this path correct?

Author: Yes. `wazuh-certs-tool.sh` is creating this directory.

Reviewer: What path would it be if using custom certs?

Author: `"{{ playbook_dir }}/wazuh/custom_certificates"` for example: `etc/kayobe/ansible/wazuh/custom_certificates/root-ca.pem`

Reviewer: Is it not also present at `etc/kayobe/ansible/wazuh/certificates/root-ca.pem` when generating certs? I'm looking for a standard path to reference in both cases.

Author: When certs are autogenerated they are generated to `etc/kayobe/ansible/wazuh/certificates/wazuh-certificates/`.  Path `etc/kayobe/ansible/wazuh/certificates` is used by Wazuh playbooks to download wazuh-certs-tool.sh, and this tool is creating wazuh-certificates directory and generates certs there, later certs are copied to wazuh host. 
In custom cert scenario, certs are stored in `etc/kayobe/ansible/wazuh/custom_certificates` and later are copied to wazuh host.

You can have `etc/kayobe/ansible/wazuh/certificates/root-ca.pem` and use it for signing certs generated by `wazuh-certs-tool.sh` (tool itself is supporting it), but this is manual action, as upstream roles currently don't support this from ansible perspective. 

So tl;dr answer is: no.

As for standard path. I think it is handy to have this,  but right now we are in place where cert tool is started on every playbook run, so it strong candidate for follow up task.
